_posts/2021-10-25-clean-architecture-story.md

```diff
@@ -0,0 +1,332 @@
+we are safe, that any other interaction, any other use case, is not going to break these rules. This approach also
+simplifies testing, as business rules can be verified in separation from the use case interaction logic.
+
+## Where is the “Clean Architecture”?
+Let‘s stop with business logic for a moment. We created quite thoughtful, extensible code for sure, but why are we
+talking about “Clean” architecture? We already used Domain-Driven Design and Hexagonal architecture concepts. Is there
+anything more? Imagine that another person is going to help us with implementation. She is not aware of the source code
+yet and simply would like to take a look at the codebase. And she sees: ![Use case classes](/img/articles/2021-10-25-clean-architecture-story/use_case_classes.png)
+It looks like something to her, doesn‘t it? A kind of reservation system! It is not yet another domain service with
+some methods that have no clear connection with possible uses — the list of classes itself describes what the system
+can do.
+
+## The first assumption
+We have a mocked implementation as the schedule creator. It is OK to test logic at the unit test level, but not enough
+to run a prototype.
+
+After a short call with our customer we know more about the daily schedule — there are six slots, two hours each,
+starting at 8:oo a.m. We also know that this recipe for the daily schedule is very, very simple and it is going to be
+changed soon (e.g. to accommodate for holidays, etc.). All these issues will be solved later, now we are in the
+prototype stage and our desired outcome is to have a working demo for our stranger.
+
+Where to put this simple implementation of the schedule creator?So far, the domain used an interface for that. Are we
+going to put an implementation of this interface to the infrastructure package and treat it as something outside the
+domain? Certainly not! It is simple but this is part of the domain itself, we simply replace the interface with class
+specification.
+
+```kotlin
+package eu.kowalcze.michal.arch.clean.example.domain.model
+
+class DayScheduleCreator {
+    fun create(scheduleDay: LocalDate): DaySchedule = DaySchedule(
+        scheduleDay,
+        createStandardSlots()
+    )
+//...
+}
+```
+
+(full commit: [GitHub](https://github.com/michal-kowalcze/clean-architecture-example/commit/2792fc31e98d76a610561636f03073dee73fbb47))
+
+## The prototype
+I will not be original here — for the first prototype version the REST API sounds like something reasonable. Do we care
+about other infrastructure at the moment? Persistence? No! In the previous commits a map-based persistence layer is
+used for unit tests and this solution is good enough to start with. As long as the system is not restarted, of course.
+
+What is important at this stage? We are introducing an A-P-I — this is a separate layer, so it is crucial to ensure
```

Comment:
Skąd taka pisownia `A-P-I`?

Reply:
Miałem chęć innego wypunktowania. Tak, jak czytam po dystansie czasowym to nie ma takiej potrzeby, bold wystarczy.